File: 백준/Gold/1068. 트리/트리.py
import sys

n = int(sys.stdin.readline().rstrip())
cnt = 0
tree = [[] for _ in range(n)]
visited = [False] * n
root = -1


def dfs(node):
    global cnt

    visited[node] = True

    if not tree[node]:
        cnt += 1
        return

    for child in tree[node]:
        if not visited[child]:
            dfs(child)


def remove_node(node):
    visited[node] = True

    for parent_node in tree:
        if node in parent_node:
            parent_node.remove(node)

    for child in tree[node]:
        if not visited[child]:
            remove_node(child)

    tree[node].clear()


parents = list(map(int, sys.stdin.readline().split(" ")))

for i in range(n):
    if parents[i] == -1:
        root = i
    else:
        tree[parents[i]].append(i)

removeNode = int(sys.stdin.readline().rstrip())

if removeNode == root:
    print(0)
    sys.exit()

remove_node(removeNode)
dfs(root)
print(cnt)

File: 백준/Gold/1068. 트리/트리.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
	static ArrayList<Integer>[] tree;
	static boolean[] visited;
	static int count = 0;

	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt(br.readLine());
		tree = new ArrayList[n];
		visited = new boolean[n];

		for (int i = 0; i < n; i++) {
			tree[i] = new ArrayList<>();
		}

		StringTokenizer st = new StringTokenizer(br.readLine());
		int root = -1;

		for (int i = 0; i < n; i++) {
			int parent = Integer.parseInt(st.nextToken());
			if (parent == -1) {
				root = i;
			} else {
				tree[parent].add(i);
			}
		}

		int removeNode = Integer.parseInt(br.readLine());
		if (removeNode == root) {
			System.out.println(0);
			return;
		}

		removeNode(removeNode);
		dfs(root);
		System.out.println(count);
	}

	private static void  dfs(int node) {
		visited[node] = true;

		if (tree[node].isEmpty()) {
			count++;
			return;
		}

		for (int child : tree[node]) {
			if (!visited[child]) {
				dfs(child);
			}
		}
	}

	private static void removeNode(int node) {
		visited[node] = true;

		for(ArrayList<Integer> parentNode : tree) {
			parentNode.remove((Integer) node);
		}

		for (int child : tree[node]) {
			if (!visited[child]) {
				removeNode(child);
			}
		}

		tree[node].clear();
	}
}
